### PR TITLE
Add SharedBitmappedBlock

### PR DESCRIPTION
**Pull request by jercaianu**
Reopened the PR here https://github.com/dlang/phobos/pull/6206 (which I brutally destroyed)

### CODE REVIEW

**Comment by wilzbach** [comment]
Still needs a changelog entry ;-)

File: changelog/shared-bitmapped-block.dd [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Added the SharedBitmappedBlock, as the threadsafe version of the regular BitmappedBlock.
+
+The limitation of the new implementation is that the shared version can only allocate 
+at most one block per allocation. This however allows for greater performance, because
+the synchronization is done at a single block level, whereas for multiple blocks, the whole
+bit array should have been synchronized by means of wide locking.
+
+
```
Comment (wilzbach) [inline changelog/shared-bitmapped-block.dd]: - Empty lines at the end
- Link the new symbol once in the first sentence: `$(REF SharedBitmappedBlock, std,experimental,allocator,building_blocks,bitmapped_block)`

File: changelog/shared-bitmapped-block.dd [hunk 2]
```diff
@@ -0,0 +1,8 @@
+Added the SharedBitmappedBlock, as the threadsafe version of the regular BitmappedBlock.
+
+The limitation of the new implementation is that the shared version can only allocate 
+at most one block per allocation. This however allows for greater performance, because
+the synchronization is done at a single block level, whereas for multiple blocks, the whole
```
Comment (wilzbach) [inline changelog/shared-bitmapped-block.dd]: s/ whereas for multiple blocks, the/ whereas for multiple blocks the/

**Comment by jercaianu** [comment]
@wilzbach Thanks for comments!

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 3]
```diff
@@ -1746,3 +1760,334 @@ private struct BitVector
...
+        static assert(parentAlignment >= ulong.alignof);
+    }
+
+    /**
+    The _alignment offered is user-configurable statically through parameter
+    `theAlignment`, defaulted to `platformAlignment`.
+    */
+    alias alignment = theAlignment;
+
+    // state {
```
Comment (wilzbach) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: This never closes

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 4]
```diff
@@ -1746,3 +1760,334 @@ private struct BitVector
...
+        }
+        _blocks = cast(typeof(_blocks)) localBlocks;
+    }
+
+    /// Ditto
+    static if (chooseAtRuntime == theBlockSize)
+    this(ubyte[] data, uint blockSize)
+    {
+        this._blockSize = blockSize;
+        this(data);
```
Comment (wilzbach) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: Uncovered.

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 5]
```diff
@@ -1746,3 +1760,334 @@ private struct BitVector
...
+        size_t localBlocks;
+        for (localBlocks = _blocks; localBlocks; --localBlocks)
+        {
+            immutable controlWords = localBlocks.divideRoundUp(64);
+            auto payload = data[controlWords * 8 .. $].roundStartToMultipleOf(
+                alignment);
+            if (payload.length < localBlocks * blockSize)
+            {
+                // Overestimated
+                continue;
```
Comment (wilzbach) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: Uncovered.

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 6]
```diff
@@ -1746,3 +1760,334 @@ private struct BitVector
...
+    private size_t totalAllocation(size_t capacity)
+    {
+        auto blocks = capacity.divideRoundUp(blockSize);
+        auto leadingUlongs = blocks.divideRoundUp(64);
+        import std.algorithm.comparison : min;
+        immutable initialAlignment = min(parentAlignment,
+            1U << trailingZeros(leadingUlongs * 8));
+        auto maxSlack = alignment <= initialAlignment
+            ? 0
+            : alignment - initialAlignment;
```
Comment (wilzbach) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: Uncovered.

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 7]
```diff
@@ -1746,3 +1760,334 @@ private struct BitVector
...
+    {
+        alias blockSize = theBlockSize;
+    }
+    else
+    {
+        @property uint blockSize() { return _blockSize; }
+        @property void blockSize(uint s)
+        {
+            assert(_control.length == 0 && s % alignment == 0);
+            _blockSize = s;
```
Comment (wilzbach) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: Uncovered.

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 8]
```diff
@@ -1746,3 +1760,334 @@ private struct BitVector
...
+    }
+
+    /**
+    Constructs a block allocator given a hunk of memory, or a desired capacity
+    in bytes.
+
+    $(UL
+    $(LI If `ParentAllocator` is $(REF_ALTTEXT `NullAllocator`, NullAllocator, std,experimental,allocator,building_blocks,null_allocator),
+    only the constructor taking `data` is defined and the user is responsible for freeing `data` if desired.)
+    $(LI Otherwise, both constructors are defined. The `data`-based
```
Comment (wilzbach) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: You got four:

image

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 9]
```diff
@@ -1746,3 +1760,334 @@ private struct BitVector
...
+    constructor assumes memory has been allocated with the parent allocator.
```
Comment (wilzbach) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: assumes that the memory

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 10]
```diff
@@ -1746,3 +1760,334 @@ private struct BitVector
...
+        return leadingUlongs * 8 + maxSlack + blockSize * blocks;
+    }
+
+    /**
+    Constructs a block allocator given a hunk of memory, or a desired capacity
+    in bytes.
+
+    $(UL
+    $(LI If `ParentAllocator` is $(REF_ALTTEXT `NullAllocator`, NullAllocator, std,experimental,allocator,building_blocks,null_allocator),
+    only the constructor taking `data` is defined and the user is responsible for freeing `data` if desired.)
```
Comment (wilzbach) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: Why would anyone want to use this with the NullAllocator?

File: changelog/shared-bitmapped-block.dd [hunk 11]
```diff
@@ -0,0 +1,7 @@
+Added the `SharedBitmappedBlock`, as the thread-safe version of the regular `BitmappedBlock`.
+
+The limitation of the new $(REF SharedBitmappedBlock, std,experimental,allocator,building_blocks,bitmapped_block)
+implementation is that the shared version can only allocate
+at most one block per allocation. This however allows for greater performance, because
```
Comment (wilzbach) [inline changelog/shared-bitmapped-block.dd]: This, however, allows

File: changelog/shared-bitmapped-block.dd [hunk 12]
```diff
@@ -0,0 +1,7 @@
...
+the synchronization is done at a single block level, whereas for multiple blocks the whole
+bit array should have been synchronized by means of wide locking.
```
Comment (wilzbach) [inline changelog/shared-bitmapped-block.dd]: would require synchronized access by ... ?

**Comment by jercaianu** [comment]
I closed this PR in favor of [1], which includes this functionality
https://github.com/dlang/phobos/pull/6408

**Pull request closed without merging**